### PR TITLE
CORE-17493: Add cordappName field to contract and workflow fields of cpk2

### PR DESCRIPTION
**Pull request by malachyb**
In a cordapp CPK we have a field that sets a name for a cordapp that persists through versions. However, this is assigned automatically based on other values, so we are adding the ability for users to set this name for their own cordapp manually to allow for smoother upgrades where those other values may need to change, as well as to help make CPKs more identifiable to developers.

### CODE REVIEW

**Review by kyriathar — COMMENTED** [review]
Looks good in principle. A few comments.

File: cordapp-cpk2/README.md
Comment (kyriathar) [inline cordapp-cpk2/README.md:72]: Should be adding a few words to mention what `cordappName` is for?  I believe it will raise confusion with `name`.

**Comment by simon-johnson-r3** [comment]
We should hold off merging this until Corda 5.1 is released.

File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java
Comment (simon-johnson-r3) [reply cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java:545]: This code (what was here originally) looks a little peculiar. It's saying if there was a workflow use that as the bundle description, otherwise use the contract. I wasn't aware you could define both in the same bundle. Contract bundles are loaded into verification sandboxes and workflow bundles are loaded into flow sandboxes. As far as I know we shouldn't see both kinds of classes in the same bundle at the same time.

I kind of assumed when looking at the README if you defined both in the same module you got 2 bundles.

I also see this function is called `configureCordappAttributes` which suggests it's setting these at app scope and not CPK scope.

Comment (simon-johnson-r3) [reply cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java:545]: Ok, having unpacked a CPB it looks as I'd expect, i.e. CPB contains only
```
Corda-CPB-Format: 2.0
Corda-CPB-Name: packaging-verification-app-v1
Corda-CPB-Version: 5.0.0.0-SNAPSHOT
```

And each CPK (contract and workflow) jar contains (amongst other things):
```
Corda-CPK-Cordapp-Name: net.corda.packaging-verification-app-v1
```
and
```
Corda-CPK-Cordapp-Name: net.corda.packaging-verification-contract-v1
```

So why the plugin suggests these are cordapp wide attributes is beyond me.

Comment (simon-johnson-r3) [reply cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java:545]: For now I'd just remove the `if (!setName)` check, and have it overwrite. At least then it's consistent with the other attributes. Then you could roll the `attributes.put(CPK_CORDAPP_NAME, symbolicName);` part into the other blocks as @kyriathar suggests below.

File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java
Comment (kyriathar) [inline cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java:520]:
```suggestion
            if (cordappCpkName != null)) {
```

File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java
Comment (kyriathar) [inline cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java:546]:
```suggestion
            if (cordappCpkName != null) {
```

File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java
Comment (kyriathar) [inline cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java:547]: Different pattern applies to workflow's CPK cordapp name compared to contract?

File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java
Comment (kyriathar) [inline cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java:551]: I believe this exception message is missing a "and dashes."

**Review by kyriathar — APPROVED** [review]
LGTM

**Review by simon-johnson-r3 — APPROVED** [review]

**Review by simon-johnson-r3 — CHANGES_REQUESTED** [review]
this looks fine, but I've blocked it whilst we assess what the consequence of merging it is - which we will do on slack

File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java
Comment (kyriathar) [inline cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java:518]: Is this now being used?

File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java
Comment (kyriathar) [inline cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappPlugin.java:525]: Think we could also bake into the function this piece? Then it becomes something like `populateCordappCpkName`

**Review by kyriathar — DISMISSED** [review]
LGTM, but let's wait for https://github.com/corda/corda-gradle-plugins/pull/596 PR to be merged first to this one.

**Comment by kyriathar** [comment]
Also before merging we need to update _changelog.md_ file with the changed introduced. Had a chat with @simon-johnson-r3 and we said let's add these changes to v7.0.4

File: changelog.md
Comment (malachyb) [reply changelog.md:10]: I thought it was a relevant thing to include since it also noted the Bndlib upgrade but if you think it's unnecessary I can remove it.

Comment (simon-johnson-r3) [reply changelog.md:10]: I think it's ok to add it. I don't imagine we've ever set hard rules for what goes in here.

**Review by simon-johnson-r3 — APPROVED** [review]

**Review by kyriathar — APPROVED** [review]

**Pull request merged**
